packages/template-retail-react-app/app/components/island/index.jsx

```diff
@@ -0,0 +1,222 @@
+                // `display="contents"` is problematic. The workaround is to apply the `IntersectionObserver`
+                // to the so far empty container directly. But for this we need to change its `display`
+                // settings.
+                element = ref.current
+                element.style.display = 'inline-block'
+            }
+            const observer = new IntersectionObserver(
+                (entries) => {
+                    for (const entry of entries) {
+                        if (entry.isIntersecting || entry.intersectionRatio > 0) {
+                            observer.unobserve(element)
+                            if (element === ref.current) {
+                                // Special case where `clientOnly` is `true`. Here we need to reset the `display`
+                                // settings we just had to change.
+                                element.style.display = 'contents'
+                            }
+                            hydrate()
+                            break // break on first match
+                        }
+                    }
+                },
+                typeof options === 'object' && options !== null
+                    ? options
+                    : {
+                          rootMargin: '250px'
+                      }
+            )
+            observer.observe(element)
+            return () => observer.disconnect()
+        }
+
+        // `hydrateOn='load'`
+        hydrate()
+    }, [hydrateOn, hydrated])
+
+    if (hydrated) {
+        if (!ssr && !context && ref.current) {
+            // Before finally hydrating a root-level context on the client, store a reference to the original
+            // server-rendered HTML to be consumed by eventually client-rendered nested/child islands. Unfortunately,
+            // there's no better way to reference the elements than an active DOM query. React does not yet know the
+            // child elements at this specific point in time, and the child DOM tree only exists at all because of
+            // a previous SSR step.
+            const islands = findChildren(children, Island)
+            const elements = Array.from(ref.current.querySelectorAll('[data-sfdc-island]'))
+            for (let i = 0; i < islands.length; i++) {
+                const island = islands.at(i)
+                const el = elements.at(i)
+                island && el && data.set(islands.at(i)?.props, el?.innerHTML)
+            }
+        }
+
+        return (
+            <IslandContext.Provider value={{hydrated, data}}>
+                <div
+                    ref={ref}
+                    style={{display: 'contents'}}
+                    data-sfdc-island
+                    data-sfdc-island-origin={ssr ? 'server' : 'client'}
+                >
+                    {ssr && clientOnly ? null : children}
+                </div>
+            </IslandContext.Provider>
+        )
+    }
```

Comment:
Checking my understanding...

This `if` statement does not run on every `render` right after hydration right? I think it only runs once before hydration is completed?